src/debugger/Debugger.cpp

```diff
@@ -229,6 +247,100 @@ bool Debugger::processIncomingMessages(ExecutionState* state)
     return enabled();
 }
 
+void Debugger::getBacktrace(ExecutionState* state, uint32_t minDepth, uint32_t maxDepth, bool getTotal)
+{
+    StringBuilder builder;
+    uint32_t total = 0;
+
+    while (state) {
+        FunctionObject* callee = state->resolveCallee();
+
+        while (state) {
+            if (!state->lexicalEnvironment() || !state->lexicalEnvironment()->record()) {
+                break;
+            }
+            if (state->lexicalEnvironment()->record()->isGlobalEnvironmentRecord()) {
+                break;
+            } else if (state->lexicalEnvironment()->record()->isDeclarativeEnvironmentRecord() && state->lexicalEnvironment()->record()->asDeclarativeEnvironmentRecord()->isFunctionEnvironmentRecord()) {
+                break;
+            }
+            state = state->parent();
+        }
+
+        total++;
+        if (total <= minDepth) {
+            state = state->parent();
+            continue;
+        }
+
+        if (maxDepth > 0 && total > maxDepth) {
+            if (!getTotal) {
+                break;
+            }
+
+            state = state->parent();
+            continue;
+        }
+
+        ExtendedNodeLOC loc(0, 0, SIZE_MAX);
+        loc.actualCodeBlock = nullptr;
+        String* src = nullptr;
+
+        if (!callee && state && state->lexicalEnvironment()) {
```

Comment:
Most of this code is copied from Sandbox.cpp, and I am not sure how correct it is. It works on simple examples though.

Reply:
It is correct.
How about to use same code SandBox.cpp and Debugger.cpp?

Reply:
I can, but I would need some suggestion how to do it. If I understand correctly, the current code builds the trace in `m_stackTraceData` and only when an exception happens:
https://github.com/Samsung/escargot/blob/master/src/runtime/SandBox.cpp#L104

When stack trace is requested by the debugger, what would you prefer?
1) Shall we fill `m_stackTraceData` and clear it somehow?
2) Shall we create a separate function which receives a `StackTraceDataVector&` and fills it?
3) Call `throwException()` with a dummy value, capture it by the deugger, and get stack trace?
4) Any other idea?

Reply:
I prefer second opinion you gave.
but you can do anything what you want